packages/core/test/unit/patchers/aws3_p.test.js

```diff
@@ -0,0 +1,239 @@
+var assert = require('chai').assert;
+var chai = require('chai');
+var EventEmitter = require('events');
+var sinon = require('sinon');
+var sinonChai = require('sinon-chai');
+var util = require('util');
+
+var Aws = require('../../../lib/segments/attributes/aws');
+var awsPatcher = require('../../../lib/patchers/aws3_p');
+var contextUtils = require('../../../lib/context_utils');
+var Segment = require('../../../lib/segments/segment');
+var Utils = require('../../../lib/utils');
+
+var { MiddlewareStack } = require('@aws-sdk/middleware-stack');
+
+var logger = require('../../../lib/logger').getLogger();
+
+chai.should();
+chai.use(sinonChai);
+
+var traceId = '1-57fbe041-2c7ad569f5d6ff149137be86';
+
+describe('AWS v3 patcher', function() {
```

Comment:
In the past we've had an issue where we'd create subsegments for presigning S3 URLs, even though it's a local request. But I'm *pretty sure* this shouldn't be a problem with the new V3 design, since the S3 presigner is its [own package](https://github.com/aws/aws-sdk-js-v3/tree/master/packages/s3-request-presigner) rather than implemented as a normal S3 client operation like before. Can we add a unit test to make sure that presigns won't get captured?

Reply:
Good catch! It's definitely a problem because the presign packages add themselves as a middleware. I'll look into it!

Reply:
I see. If we just add our middleware before theirs using the relative ordering would that fix it? If it's a middleware as opposed to a client can it get captured?

Reply:
Unfortunately, the presign middleware inserts itself right before the `AwsAuthMiddleware` (https://github.com/aws/aws-sdk-js-v3/blob/master/packages/s3-request-presigner/src/getSignedUrl.ts#L46) so we cannot get in between anymore. The most straight forward implementation is to leave it up to the user. If they want to have the presigned URL instrumentalized (which I think is a feature not a bug) then they can add the XRay patcher.

```javascript
import { getSignedUrl } from "@aws-sdk/s3-request-presigner";
import { S3Client, GetObjectCommand } from "@aws-sdk/client-s3";
const client = new S3Client(clientParams); // feel free to wrap in captureAWSClient()
const command = new GetObjectCommand(getObjectParams);
const url = await getSignedUrl(client, command, { expiresIn: 3600 });
```

Reply:
I'm a little confused, if we're instrumenting by replacing the `client.send()` function does `getSignedUrl` use that under the hood? Right now, we'll only be adding our middleware which adds a trace header which won't be actually used for anything. I don't think an actual subsegment will be made just for `getSignedUrl`.

Reply:
Yes, `getSignedUrl` calls `send` here https://github.com/aws/aws-sdk-js-v3/blob/master/packages/s3-request-presigner/src/getSignedUrl.ts#L51.
I also noticed that `getSignedUrl` removes headers from some other middlewares. I wonder if we could convince them to include the XRay header as well (https://github.com/aws/aws-sdk-js-v3/blob/master/packages/s3-request-presigner/src/getSignedUrl.ts#L26).
Like I said earlier, giving the developer the choice to instrumentalize a presigned request could be a feature rather than a bug. I'd make sure this behavior is well documented and let the developer choose. They can always create a new S3 client without the xray middleware applied when they presign URLs.

Reply:
Ok thank you for pointing me to those links! I'd like to try this out and get a feel for it myself, but I realized that this API hasn't been exposed on your branch yet. Can you export it from our top-level `aws-xray` module and include an [automatic](https://github.com/aws/aws-xray-sdk-node/tree/master/packages/core#capture-outgoing-aws-requests-on-a-single-client) and [manual]() mode example in the readme like the ones I linked for V2? We already have the `captureAWSClient` API for v2 clients, so you'll need to change the function to something like `captureV3AWSClient`.

After I do some testing and we figure out the `getSignedUrl` experience we want, this should be good to go. Thanks for bearing with us!

Reply:
I have another quick question/consideration.

If all v3 clients implement packages/types/src/client.ts, is it necessary to have an outward-facing 'v3' version of the captureAWSClient method?  That is, could captureAWSClient check for the client interface implementation in the given argument, and forward the client to a "private" captureAWSv3Client method if the check passes?

(this message will also self-destruct)

Reply:
After testing this I think capturing `getSignedUrl` as we are currently is the correct experience. I think I mainly misremembered previous customer issues that arose from presigining behavior thinking they wanted it removed, but that was not the case. Thanks for providing your input here!